File: shims/qpid-proton-cpp/src/qpidit/jms_hdrs_props_test/Sender.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

#include "qpidit/jms_hdrs_props_test/Sender.hpp"
#include "qpidit/Base64.hpp"

#include <cerrno>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <json/json.h>
#include <proton/connection.hpp>
#include <proton/container.hpp>
#include <proton/thread_safe.hpp>
#include <proton/tracker.hpp>
#include <proton/transport.hpp>
#include <stdio.h>

namespace qpidit
{
    namespace jms_hdrs_props_test
    {
        Sender::Sender(const std::string& brokerUrl,
                       const std::string& jmsMessageType,
                       const Json::Value& testParams) :
                _brokerUrl(brokerUrl),
                _jmsMessageType(jmsMessageType),
                _testValueMap(testParams[0]),
                _testHeadersMap(testParams[1]),
                _testPropertiesMap(testParams[2]),
                _msgsSent(0),
                _msgsConfirmed(0),
                _totalMsgs(getTotalNumMessages(_testValueMap))
        {
            if (_testValueMap.type() != Json::objectValue) {
                throw qpidit::InvalidJsonRootNodeError(Json::objectValue, _testValueMap.type());
            }
        }

        Sender::~Sender() {}

        void Sender::on_container_start(proton::container &c) {
            c.open_sender(_brokerUrl);
        }

        void Sender::on_sendable(proton::sender &s) {
            if (_totalMsgs == 0) {
                s.connection().close();
            } else if (_msgsSent == 0) {
                Json::Value::Members subTypes = _testValueMap.getMemberNames();
                std::sort(subTypes.begin(), subTypes.end());
                for (std::vector<std::string>::const_iterator i=subTypes.begin(); i!=subTypes.end(); ++i) {
                    sendMessages(s, *i, _testValueMap[*i]);
                }
            }
        }

        void Sender::on_tracker_accept(proton::tracker &t) {
            _msgsConfirmed++;
            if (_msgsConfirmed == _totalMsgs) {
                t.connection().close();
            }
        }

        void Sender::on_transport_close(proton::transport &t) {
            _msgsSent = _msgsConfirmed;
        }

        // protected

        void Sender::sendMessages(proton::sender &s, const std::string& subType, const Json::Value& testValues) {
            uint32_t valueNumber = 0;
            for (Json::Value::const_iterator i=testValues.begin(); i!=testValues.end(); ++i) {
                if (s.credit()) {
                    proton::message msg;
                    if (_jmsMessageType.compare("JMS_MESSAGE_TYPE") == 0) {
                        setMessage(msg, subType, (*i).asString());
                    } else if (_jmsMessageType.compare("JMS_BYTESMESSAGE_TYPE") == 0) {
                        setBytesMessage(msg, subType, (*i).asString());
                    } else if (_jmsMessageType.compare("JMS_MAPMESSAGE_TYPE") == 0) {
                        setMapMessage(msg, subType, (*i).asString(), valueNumber);
                    } else if (_jmsMessageType.compare("JMS_OBJECTMESSAGE_TYPE") == 0) {
                        setObjectMessage(msg, subType, *i);
                    } else if (_jmsMessageType.compare("JMS_STREAMMESSAGE_TYPE") == 0) {
                        setStreamMessage(msg, subType, (*i).asString());
                    } else if (_jmsMessageType.compare("JMS_TEXTMESSAGE_TYPE") == 0) {
                        setTextMessage(msg, *i);
                    } else {
                        throw qpidit::UnknownJmsMessageTypeError(_jmsMessageType);
                    }
                    addMessageHeaders(msg);
                    addMessageProperties(msg);
                    s.send(msg);
                    _msgsSent += 1;
                    valueNumber += 1;
                }
            }

        }

        proton::message& Sender::setMessage(proton::message& msg, const std::string& subType, const std::string& testValueStr) {
            if (subType.compare("none") != 0) {
                throw qpidit::UnknownJmsMessageSubTypeError(subType);
            }
            if (testValueStr.size() != 0) {
                throw InvalidTestValueError(subType, testValueStr);
            }
            msg.content_type(proton::symbol("application/octet-stream"));
            msg.message_annotations().put(proton::symbol("x-opt-jms-msg-type"), s_jmsMessageTypeAnnotationValues["JMS_MESSAGE_TYPE"]);
            return msg;
        }

        proton::message& Sender::setBytesMessage(proton::message& msg, const std::string& subType, const std::string& testValueStr) {
            proton::binary bin;
            if (subType.compare("boolean") == 0) {
                if (testValueStr.compare("False") == 0) bin.push_back(char(0));
                else if (testValueStr.compare("True") == 0) bin.push_back(char(1));
                else throw InvalidTestValueError(subType, testValueStr);
            } else if (subType.compare("byte") == 0) {
                uint8_t val = getIntegralValue<int8_t>(testValueStr);
                bin.push_back(char(val));
            } else if (subType.compare("bytes") == 0) {
                bin = b64_decode(testValueStr);
            } else if (subType.compare("char") == 0) {
                std::string decodedStr = b64_decode(testValueStr);
                bin.push_back(char(0));
                if (testValueStr[0] == '\\') { // Format: '\xNN'
                    bin.push_back(getIntegralValue<char>(decodedStr.substr(2)));
                } else { // Format: 'c'
                    bin.push_back(decodedStr[0]);
                }
            } else if (subType.compare("double") == 0) {
                uint64_t val;
                try {
                    val = htobe64(std::strtoul(testValueStr.data(), NULL, 16));
                } catch (const std::exception& e) { throw qpidit::InvalidTestValueError("double", testValueStr); }
                numToBinary(val, bin);
               //for (int i=0; i<sizeof(val); ++i) {
               //     bin.push_back(* ((char*)&val + i));
               // }
            } else if (subType.compare("float") == 0) {
                uint32_t val;
                try {
                    val = htobe32((uint32_t)std::strtoul(testValueStr.data(), NULL, 16));
                } catch (const std::exception& e) { throw qpidit::InvalidTestValueError("float", testValueStr); }
                numToBinary(val, bin);
                //for (int i=0; i<sizeof(val); ++i) {
                //    bin.push_back(* ((char*)&val + i));
                //}
            } else if (subType.compare("long") == 0) {
                uint64_t val = htobe64(getIntegralValue<uint64_t>(testValueStr));
                numToBinary(val, bin);
                //bin.assign(sizeof(val), val);
            } else if (subType.compare("int") == 0) {
                uint32_t val = htobe32(getIntegralValue<uint32_t>(testValueStr));
                numToBinary(val, bin);
                //bin.assign(sizeof(val), val);
            } else if (subType.compare("short") == 0) {
                uint16_t val = htobe16(getIntegralValue<int16_t>(testValueStr));
                numToBinary(val, bin);
                //bin.assign(sizeof(val), val);
            } else if (subType.compare("string") == 0) {
                std::ostringstream oss;
                uint16_t strlen = htobe16((uint16_t)testValueStr.size());
                oss.write((char*)&strlen, sizeof(strlen));
                oss << testValueStr;
                std::string os = oss.str();
                bin.assign(os.begin(), os.end());
            } else {
                throw qpidit::UnknownJmsMessageSubTypeError(subType);
            }
            msg.body(bin);
            msg.inferred(true);
            msg.content_type(proton::symbol("application/octet-stream"));
            msg.message_annotations().put(proton::symbol("x-opt-jms-msg-type"), s_jmsMessageTypeAnnotationValues["JMS_BYTESMESSAGE_TYPE"]);
            return msg;
        }

        proton::message& Sender::setMapMessage(proton::message& msg, const std::string& subType, const std::string& testValueStr, uint32_t valueNumber) {
            std::ostringstream oss;
            oss << subType << std::setw(3) << std::setfill('0') << valueNumber;
            std::string mapKey(oss.str());
            std::map<std::string, proton::value> m;
            if (subType.compare("boolean") == 0) {
                if (testValueStr.compare("False") == 0) m[mapKey] = false;
                else if (testValueStr.compare("True") == 0) m[mapKey] = true;
                else throw InvalidTestValueError(subType, testValueStr);
            } else if (subType.compare("byte") == 0) {
                m[mapKey] = int8_t(getIntegralValue<int8_t>(testValueStr));
            } else if (subType.compare("bytes") == 0) {
                m[mapKey] = b64_decode(testValueStr);
            } else if (subType.compare("char") == 0) {
                std::string decodedStr = b64_decode(testValueStr);
                wchar_t val;
                if (decodedStr[0] == '\\') { // Format: '\xNN'
                    val = (wchar_t)getIntegralValue<wchar_t>(decodedStr.substr(2));
                } else { // Format: 'c'
                    val = decodedStr[0];
                }
                m[mapKey] = val;
            } else if (subType.compare("double") == 0) {
                m[mapKey] = getFloatValue<double, uint64_t>(testValueStr);
            } else if (subType.compare("float") == 0) {
                m[mapKey] = getFloatValue<float, uint32_t>(testValueStr);
            } else if (subType.compare("int") == 0) {
                m[mapKey] = getIntegralValue<int32_t>(testValueStr);
            } else if (subType.compare("long") == 0) {
                m[mapKey] = getIntegralValue<int64_t>(testValueStr);
            } else if (subType.compare("short") == 0) {
                m[mapKey] = getIntegralValue<int16_t>(testValueStr);
            } else if (subType.compare("string") == 0) {
                m[mapKey] = testValueStr;
            } else {
                throw qpidit::UnknownJmsMessageSubTypeError(subType);
            }
            msg.inferred(false);
            msg.body(m);
            msg.message_annotations().put(proton::symbol("x-opt-jms-msg-type"), s_jmsMessageTypeAnnotationValues["JMS_MAPMESSAGE_TYPE"]);
            return msg;
        }

        proton::message& Sender::setObjectMessage(proton::message& msg, const std::string& subType, const Json::Value& testValue) {
            msg.body(getJavaObjectBinary(subType, testValue.asString()));
            msg.inferred(true);
            msg.content_type(proton::symbol("application/x-java-serialized-object"));
            msg.message_annotations().put(proton::symbol("x-opt-jms-msg-type"), s_jmsMessageTypeAnnotationValues["JMS_OBJECTMESSAGE_TYPE"]);
            return msg;
        }

        proton::message& Sender::setStreamMessage(proton::message& msg, const std::string& subType, const std::string& testValueStr) {
            std::vector<proton::value> l;
            if (subType.compare("boolean") == 0) {
                if (testValueStr.compare("False") == 0) l.push_back(false);
                else if (testValueStr.compare("True") == 0) l.push_back(true);
                else throw InvalidTestValueError(subType, testValueStr);
            } else if (subType.compare("byte") == 0) {
                l.push_back(int8_t(getIntegralValue<int8_t>(testValueStr)));
            } else if (subType.compare("bytes") == 0) {
                l.push_back(b64_decode(testValueStr));
            } else if (subType.compare("char") == 0) {
                std::string decodedStr = b64_decode(testValueStr);
                wchar_t val;
                if (decodedStr[0] == '\\') { // Format: '\xNN'
                    val = (wchar_t)getIntegralValue<wchar_t>(decodedStr.substr(2));
                } else { // Format: 'c'
                    val = decodedStr[0];
                }
                l.push_back(val);
            } else if (subType.compare("double") == 0) {
                l.push_back(getFloatValue<double, uint64_t>(testValueStr));
            } else if (subType.compare("float") == 0) {
                l.push_back(getFloatValue<float, uint32_t>(testValueStr));
            } else if (subType.compare("int") == 0) {
                l.push_back(getIntegralValue<int32_t>(testValueStr));
            } else if (subType.compare("long") == 0) {
                l.push_back(getIntegralValue<int64_t>(testValueStr));
            } else if (subType.compare("short") == 0) {
                l.push_back(getIntegralValue<int16_t>(testValueStr));
            } else if (subType.compare("string") == 0) {
                l.push_back(testValueStr);
            } else {
                throw qpidit::UnknownJmsMessageSubTypeError(subType);
            }
            msg.body(l);
            msg.inferred(true);
            msg.message_annotations().put(proton::symbol("x-opt-jms-msg-type"), s_jmsMessageTypeAnnotationValues["JMS_STREAMMESSAGE_TYPE"]);
            return msg;
       }

        proton::message& Sender::setTextMessage(proton::message& msg, const Json::Value& testValue) {
            msg.body(testValue.asString());
            msg.inferred(false);
            msg.message_annotations().put(proton::symbol("x-opt-jms-msg-type"), s_jmsMessageTypeAnnotationValues["JMS_TEXTMESSAGE_TYPE"]);
            return msg;
        }

        proton::message& Sender::addMessageHeaders(proton::message& msg) {
            Json::Value::Members headerNames = _testHeadersMap.getMemberNames();
            for (std::vector<std::string>::const_iterator i=headerNames.begin(); i!=headerNames.end(); ++i) {
                const Json::Value _subMap = _testHeadersMap[*i];
                const std::string headerValueType = _subMap.getMemberNames()[0]; // There is always only one entry in map
                std::string val = _subMap[headerValueType].asString();
                if (i->compare("JMS_TYPE_HEADER") == 0) {
                    setJmsTypeHeader(msg, val);
                } else if (i->compare("JMS_CORRELATIONID_HEADER") == 0) {
                    if (headerValueType.compare("bytes") == 0) {
                        setJmsCorrelationId(msg, b64_decode(val));
                    } else {
                        setJmsCorrelationId(msg, val);
                    }
                } else if (i->compare("JMS_REPLYTO_HEADER") == 0) {
                    setJmsReplyTo(msg, headerValueType, val);
                } else {
                    throw qpidit::UnknownJmsHeaderTypeError(*i);
                }
            }
            return msg;
        }

        //static
        proton::message& Sender::setJmsTypeHeader(proton::message& msg, const std::string& t) {
            msg.subject(t);
            return msg;
        }

        //static
        proton::message& Sender::setJmsCorrelationId(proton::message& msg, const std::string& cid) {
            proton::message_id mid(cid);
            msg.correlation_id(mid);
            msg.message_annotations().put(proton::symbol("x-opt-app-correlation-id"), true);
            return msg;
        }

        //static
        proton::message& Sender::setJmsCorrelationId(proton::message& msg, const proton::binary cid) {
            proton::message_id mid(cid);
            msg.correlation_id(cid);
            msg.message_annotations().put(proton::symbol("x-opt-app-correlation-id"), true);
            return msg;
        }

        //static
        proton::message& Sender::setJmsReplyTo(proton::message& msg, const std::string& dts, const std::string& d) {
            if (dts.compare("queue") == 0) {
                msg.reply_to(/*std::string("queue://") + */d);
                msg.message_annotations().put(proton::symbol("x-opt-jms-reply-to"), int8_t(qpidit::JMS_QUEUE));
            } else if (dts.compare("temp_queue") == 0) {
                msg.reply_to(/*std::string("queue://") + */d);
                msg.message_annotations().put(proton::symbol("x-opt-jms-reply-to"), int8_t(qpidit::JMS_TMEP_QUEUE));
            } else if (dts.compare("topic") == 0) {
                msg.reply_to(/*std::string("topic://") + */d);
                msg.message_annotations().put(proton::symbol("x-opt-jms-reply-to"), int8_t(qpidit::JMS_TOPIC));
            } else if (dts.compare("temp_topic") == 0) {
                msg.reply_to(/*std::string("topic://") + */d);
                msg.message_annotations().put(proton::symbol("x-opt-jms-reply-to"), int8_t(qpidit::JMS_TEMP_TOPIC));
            } else {
                throw qpidit::UnknownJmsDestinationTypeError(dts);
            }
            return msg;
        }

        proton::message& Sender::addMessageProperties(proton::message& msg) {
            Json::Value::Members propertyNames = _testPropertiesMap.getMemberNames();
            for (std::vector<std::string>::const_iterator i=propertyNames.begin(); i!=propertyNames.end(); ++i) {
                const Json::Value _subMap = _testPropertiesMap[*i];
                const std::string propertyValueType = _subMap.getMemberNames()[0]; // There is always only one entry in map
                std::string val = _subMap[propertyValueType].asString();
                if (propertyValueType.compare("boolean") == 0) {
                    if (val.compare("False") == 0) msg.properties().put(*i, false);
                    else if (val.compare("True") == 0) msg.properties().put(*i, true);
                    else throw InvalidTestValueError(propertyValueType, val);
                } else if (propertyValueType.compare("byte") == 0) {
                    msg.properties().put(*i, getIntegralValue<int8_t>(val));
                } else if (propertyValueType.compare("double") == 0) {
                    msg.properties().put(*i, getFloatValue<double, uint64_t>(val));
                } else if (propertyValueType.compare("float") == 0) {
                    msg.properties().put(*i, getFloatValue<float, uint64_t>(val));
                } else if (propertyValueType.compare("int") == 0) {
                    msg.properties().put(*i, getIntegralValue<int32_t>(val));
                } else if (propertyValueType.compare("long") == 0) {
                    msg.properties().put(*i, getIntegralValue<int64_t>(val));
                } else if (propertyValueType.compare("short") == 0) {
                    msg.properties().put(*i, getIntegralValue<int16_t>(val));
                } else if (propertyValueType.compare("string") == 0) {
                    msg.properties().put(*i, val);
                } else {
                    throw qpidit::UnknownJmsPropertyTypeError(propertyValueType);
                }
            }
            return msg;
        }

        //static
        proton::binary Sender::getJavaObjectBinary(const std::string& javaClassName, const std::string& valAsString) {
            proton::binary javaObjectBinary;
            char buf[1024];
            int bytesRead;
            FILE* fp = ::popen("java -cp target/JavaObjUtils.jar org.apache.qpid.interop_test.obj_util.JavaObjToBytes javaClassStr", "rb");
            if (fp == NULL) { throw qpidit::PopenError(errno); }
            do {
                bytesRead = ::fread(buf, 1, sizeof(buf), fp);
                javaObjectBinary.insert(javaObjectBinary.end(), &buf[0], &buf[bytesRead-1]);
            } while (bytesRead == sizeof(buf));
            int status = ::pclose(fp);
            if (status == -1) {
                throw qpidit::PcloseError(errno);
            }
            return javaObjectBinary;
        }

        // static
        uint32_t Sender::getTotalNumMessages(const Json::Value& testValueMap) {
            uint32_t tot = 0;
            for (Json::Value::const_iterator i = testValueMap.begin(); i != testValueMap.end(); ++i) {
                tot += (*i).size();
            }
            return tot;
        }

    } /* namespace jms_hdrs_props_test */
} /* namespace qpidit */



/*
 * --- main ---
 * Args: 1: Broker address (ip-addr:port)
 *       2: Queue name
 *       3: AMQP type
 *       4: JSON Test parameters containing 3 maps: [testValueMap, testHeadersMap, testPropertiesMap]
 */

int main(int argc, char** argv) {
    try {
        // TODO: improve arg management a little...
        if (argc != 5) {
            throw qpidit::ArgumentError("Incorrect number of arguments");
        }

        std::ostringstream oss;
        oss << argv[1] << "/" << argv[2];

        Json::Value testParams;
        Json::CharReaderBuilder builder;
        Json::CharReader* jsonReader = builder.newCharReader();
        std::string parseErrors;
        if (not jsonReader->parse(argv[4], argv[4] + ::strlen(argv[4]), &testParams, &parseErrors)) {
            throw qpidit::JsonParserError(parseErrors);
        }

        qpidit::jms_hdrs_props_test::Sender sender(oss.str(), argv[3], testParams);
        proton::container(sender).run();
    } catch (const std::exception& e) {
        std::cout << "Sender error: " << e.what() << std::endl;
    }
}

File: shims/qpid-proton-python/src/jms_messages_test/Sender.py
#!/usr/bin/env python

"""
JMS sender shim for qpid-interop-test
"""

#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

import base64
import json
import signal
import subprocess
import struct
import sys
import traceback

import proton
import proton.handlers
import proton.reactor
from qpid_interop_test.qit_errors import InteropTestError
from qpid_interop_test.qit_jms_types import create_annotation

class JmsMessagesTestSender(proton.handlers.MessagingHandler):
    """
    This shim sends JMS messages of a particular JMS message type according to the test parameters list. This list
    contains three maps:
    0: The test value map, which contains test value types as keys, and lists of values of that type;
    1. The test headers map, which contains the JMS headers as keys and a submap conatining types and values;
    2. The test proprties map, which contains the name of the properties as keys, and a submap containing types
       and values
    This shim takes the combinations of the above map and creates test cases, each of which sends a single message
    with (or without) JMS headers and properties.
    """
    def __init__(self, broker_url, queue_name, jms_msg_type, test_parameters_list):
        super().__init__()
        self.broker_url = broker_url
        self.queue_name = queue_name
        self.jms_msg_type = jms_msg_type
        self.test_value_map = test_parameters_list
        self.sent = 0
        self.confirmed = 0
        self.total = self._get_total_num_msgs()
        signal.signal(signal.SIGINT, self.signal_handler)
        signal.signal(signal.SIGTERM, self.signal_handler)

    def on_start(self, event):
        """Event callback for when the client starts"""
        connection = event.container.connect(url=self.broker_url, sasl_enabled=False, reconnect=False)
        event.container.create_sender(connection, target=self.queue_name)

    def on_sendable(self, event):
        """Event callback for when send credit is received, allowing the sending of messages"""
        if self.sent == 0:
            # These types expect a test_values Python string representation of a map: '{type:[val, val, val], ...}'
            for sub_type in sorted(self.test_value_map.keys()):
                if self._send_test_values(event, sub_type, self.test_value_map[sub_type]):
                    return

    def on_connection_error(self, event):
        print('JmsMessagesTestSender.on_connection_error')

    def on_session_error(self, event):
        print('JmsMessagesTestSender.on_session_error')

    def on_link_error(self, event):
        print('JmsMessagesTestSender.on_link_error')

    def on_accepted(self, event):
        """Event callback for when a sent message is accepted by the broker"""
        self.confirmed += 1
        if self.confirmed == self.total:
            event.connection.close()

    def on_disconnected(self, event):
        """Event callback for when the broker disconnects with the client"""
        self.sent = self.confirmed

    def _get_total_num_msgs(self):
        """
        Calculates the total number of messages to be sent based on the message parameters received on the command-line
        """
        total = 0
        for key in self.test_value_map.keys():
            total += len(self.test_value_map[key])
        return total

    def _send_test_values(self, event, test_value_type, test_values):
        """Method which loops through recieved parameters and sends the corresponding messages"""
        value_num = 0
        for test_value in test_values:
            if event.sender.credit:
                message = self._create_message(test_value_type, test_value, value_num)
                # TODO: set message to address
                if message is not None:
                    event.sender.send(message)
                    self.sent += 1
                    value_num += 1
                else:
                    event.connection.close()
                    return True
        return False

    # TODO: Change this to return a list of messages. That way each test can return more than one message
    def _create_message(self, test_value_type, test_value, value_num):
        """Create a single message of the appropriate JMS message type"""
        if self.jms_msg_type == 'JMS_MESSAGE_TYPE':
            return self._create_jms_message(test_value_type, test_value)
        if self.jms_msg_type == 'JMS_BYTESMESSAGE_TYPE':
            return self._create_jms_bytesmessage(test_value_type, test_value)
        if self.jms_msg_type == 'JMS_MAPMESSAGE_TYPE':
            return self._create_jms_mapmessage(test_value_type, test_value, "%s%03d" % (test_value_type, value_num))
        if self.jms_msg_type == 'JMS_OBJECTMESSAGE_TYPE':
            return self._create_jms_objectmessage('%s:%s' % (test_value_type, test_value))
        if self.jms_msg_type == 'JMS_STREAMMESSAGE_TYPE':
            return self._create_jms_streammessage(test_value_type, test_value)
        if self.jms_msg_type == 'JMS_TEXTMESSAGE_TYPE':
            return self._create_jms_textmessage(test_value)
        print('JmsMessagesTestSender: Unsupported JMS message type "%s"' % self.jms_msg_type)
        return None

    def _create_jms_message(self, test_value_type, test_value):
        """Create a JMS message type (without message body)"""
        if test_value_type != 'none':
            raise InteropTestError('JmsMessagesTestSender._create_jms_message: ' \
                                   'Unknown or unsupported subtype "%s"' % test_value_type)
        if test_value is not None:
            raise InteropTestError('JmsMessagesTestSender._create_jms_message: ' \
                                   'Invalid value "%s" for subtype "%s"' % (test_value, test_value_type))
        return proton.Message(id=(self.sent+1),
                              content_type='application/octet-stream',
                              annotations=create_annotation('JMS_MESSAGE_TYPE'))

    def _create_jms_bytesmessage(self, test_value_type, test_value):
        """Create a JMS bytes message"""
        # NOTE: test_value contains all unicode strings u'...' as returned by json
        body_bytes = None
        if test_value_type == 'boolean':
            body_bytes = b'\x01' if test_value == 'True' else b'\x00'
        elif test_value_type == 'byte':
            body_bytes = struct.pack('b', int(test_value, 16))
        elif test_value_type == 'bytes':
            body_bytes = base64.b64decode(test_value)
        elif test_value_type == 'char':
            # JMS expects two-byte chars, ASCII chars can be prefixed with '\x00'
            body_bytes = b'\x00' + base64.b64decode(test_value)
        elif test_value_type in ('double', 'float'):
            body_bytes = bytes.fromhex(test_value[2:])
        elif test_value_type == 'int':
            body_bytes = struct.pack('!i', int(test_value, 16))
        elif test_value_type == 'long':
            body_bytes = struct.pack('!q', int(test_value, 16))
        elif test_value_type == 'short':
            body_bytes = struct.pack('!h', proton.short(test_value, 16))
        elif test_value_type == 'string':
            # NOTE: First two bytes must be string length
            test_value_str = str(test_value) # remove unicode
            body_bytes = struct.pack('!H', len(test_value_str)) + test_value_str.encode('utf-8')
        else:
            raise InteropTestError('JmsMessagesTestSender._create_jms_bytesmessage: ' \
                                   'Unknown or unsupported subtype "%s"' %
                                   test_value_type)
        return proton.Message(id=(self.sent+1),
                              body=body_bytes,
                              inferred=True,
                              content_type='application/octet-stream',
                              annotations=create_annotation('JMS_BYTESMESSAGE_TYPE'))

    def _create_jms_mapmessage(self, test_value_type, test_value, name):
        """Create a JMS map message"""
        if test_value_type == 'boolean':
            value = test_value == 'True'
        elif test_value_type == 'byte':
            value = proton.byte(int(test_value, 16))
        elif test_value_type == 'bytes':
            value = base64.b64decode(test_value)
        elif test_value_type == 'char':
            value = proton.char(base64.b64decode(test_value).decode('utf-8'))
        elif test_value_type == 'double':
            value = struct.unpack('!d', bytes.fromhex(test_value[2:]))[0]
        elif test_value_type == 'float':
            value = proton.float32(struct.unpack('!f', bytes.fromhex(test_value[2:]))[0])
        elif test_value_type == 'int':
            value = proton.int32(int(test_value, 16))
        elif test_value_type == 'long':
            value = int(test_value, 16)
        elif test_value_type == 'short':
            value = proton.short(int(test_value, 16))
        elif test_value_type == 'string':
            value = test_value
        else:
            raise InteropTestError('JmsMessagesTestSender._create_jms_mapmessage: ' \
                                   'Unknown or unsupported subtype "%s"' % test_value_type)
        return proton.Message(id=(self.sent+1),
                              body={name: value},
                              inferred=False,
                              annotations=create_annotation('JMS_MAPMESSAGE_TYPE'))

    def _create_jms_objectmessage(self, test_value):
        """Create a JMS object message"""
        java_binary = self._s_get_java_obj_binary(test_value)
        return proton.Message(id=(self.sent+1),
                              body=java_binary,
                              inferred=True,
                              content_type='application/x-java-serialized-object',
                              annotations=create_annotation('JMS_OBJECTMESSAGE_TYPE'))

    @staticmethod
    def _s_get_java_obj_binary(java_class_str):
        """Call external utility to create Java object and stringify it, returning the string representation"""
        out_str = subprocess.check_output(['java',
                                           '-cp',
                                           'target/JavaObjUtils.jar',
                                           'org.apache.qpid.interop_test.obj_util.JavaObjToBytes',
                                           java_class_str])
        out_str_list = out_str.split('\n')[:-1] # remove trailing \n
        if out_str_list[0] != java_class_str:
            raise InteropTestError('JmsMessagesTestSender._s_get_java_obj_binary(): ' \
                                   'Call to JavaObjToBytes failed\n%s' % out_str)
        return bytes.fromhex(out_str_list[1])

    def _create_jms_streammessage(self, test_value_type, test_value):
        """Create a JMS stream message"""
        if test_value_type == 'boolean':
            body_list = [test_value == 'True']
        elif test_value_type == 'byte':
            body_list = [proton.byte(int(test_value, 16))]
        elif test_value_type == 'bytes':
            body_list = [base64.b64decode(test_value)]
        elif test_value_type == 'char':
            body_list = [proton.char(base64.b64decode(test_value).decode('utf-8'))]
        elif test_value_type == 'double':
            body_list = [struct.unpack('!d', bytes.fromhex(test_value[2:]))[0]]
        elif test_value_type == 'float':
            body_list = [proton.float32(struct.unpack('!f', bytes.fromhex(test_value[2:]))[0])]
        elif test_value_type == 'int':
            body_list = [proton.int32(int(test_value, 16))]
        elif test_value_type == 'long':
            body_list = [int(test_value, 16)]
        elif test_value_type == 'short':
            body_list = [proton.short(int(test_value, 16))]
        elif test_value_type == 'string':
            body_list = [test_value]
        else:
            raise InteropTestError('JmsMessagesTestSender._create_jms_streammessage: ' \
                                   'Unknown or unsupported subtype "%s"' %
                                   test_value_type)
        return proton.Message(id=(self.sent+1),
                              body=body_list,
                              inferred=True,
                              annotations=create_annotation('JMS_STREAMMESSAGE_TYPE'))

    def _create_jms_textmessage(self, test_value_text):
        """Create a JMS text message"""
        return proton.Message(id=(self.sent+1),
                              body=str(test_value_text),
                              annotations=create_annotation('JMS_TEXTMESSAGE_TYPE'))

    def on_transport_error(self, event):
        print('Sender: Broker not found at %s' % self.broker_url)

    @staticmethod
    def signal_handler(signal_number, _):
        """Signal handler"""
        if signal_number in [signal.SIGTERM, signal.SIGINT]:
            print('Receiver: received signal %d, terminating' % signal_number)
            sys.exit(1)




# --- main ---
# Args: 1: Broker address (ip-addr:port)
#       2: Queue name
#       3: JMS message type
#       4: JSON Test parameters containing 3 maps: [testValueMap, testHeadersMap, testPropertiesMap]
try:
    SENDER = JmsMessagesTestSender(sys.argv[1], sys.argv[2], sys.argv[3], json.loads(sys.argv[4]))
    proton.reactor.Container(SENDER).run()
except KeyboardInterrupt:
    pass
except Exception as exc:
    print('jms-sender-shim EXCEPTION:', exc)
    print(traceback.format_exc())
    sys.exit(1)
